File: backend/src/main/java/com/edumingle/backend/services/FollowService.java
package com.edumingle.backend.services;

import com.edumingle.backend.models.UserInfo;

import java.util.List;

public interface FollowService {
    boolean followUser(Integer followerId, Integer followedId);

    boolean unfollowUser(Integer followerId, Integer followedId);

    List<UserInfo> getFollowers(Integer userId);

    List<UserInfo> getFollowing(Integer userId);
}

File: frontend/src/app/services/follow.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { User } from '../types/user.interface';

@Injectable({
  providedIn: 'root'
})
export class FollowService {
  private baseUrl = 'http://localhost:8080/api';

  constructor(private http: HttpClient) {}

  followUser(followerId: number, followedId: number): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/follow`, { followerId, followedId });
  }

  unfollowUser(followerId: number, followedId: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/unfollow/${followerId}/${followedId}`);
  }

  getFollowers(userId: number): Observable<User[]> {
    return this.http.get<User[]>(`${this.baseUrl}/followers/${userId}`);
  }

  getFollowing(userId: number): Observable<User[]> {
    return this.http.get<User[]>(`${this.baseUrl}/following/${userId}`);
  }
}
